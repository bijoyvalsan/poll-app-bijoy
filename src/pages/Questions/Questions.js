import React, { useEffect, useCallback, useState } from "react";
import { useHistory } from "react-router-dom";
import PropTypes from "prop-types";
import {
  Question,
  PageContainer,
  Button as AddQuestion,
  Textbox
} from "../../components";
import { routes } from "../../config/routeConfig";

import {
  StyledContainer,
  StyledAddQuestionContainer
} from "./Questions.styled";

const Questions = ({ getQuestions, isLoading, questions, saveQuestion }) => {
  const [newQuestionValue, setNewQuestionValue] = useState("");
  const [newQuestionChoiceValue, setNewQuestionChoiceValue] = useState("");
  const history = useHistory();
  useEffect(() => {
    getQuestions();
  }, [getQuestions]);

  const handleOnQuestionClick = url => {
    return () => {
      history.push(`${routes.QUESTION_DETAILS}/${encodeURIComponent(url)}`);
    };
  };

  const handleOnSaveQuestionClick = useCallback(() => {
    saveQuestion({
      question: newQuestionValue,
      choices: newQuestionChoiceValue.split(",")
    });
  }, [saveQuestion, newQuestionValue, newQuestionChoiceValue]);

  const handleOnQuestionChange = useCallback(
    e => {
      setNewQuestionValue(e.target.value);
    },
    [setNewQuestionValue]
  );

  const handleOnQuestionChoiceChange = useCallback(
    e => {
      setNewQuestionChoiceValue(e.target.value);
    },
    [setNewQuestionChoiceValue]
  );
  const disableAddQuestionBtn =
    newQuestionValue === "" || newQuestionChoiceValue === "";
  return (
    <>
      <PageContainer title="Add Question">
        <StyledAddQuestionContainer>
          <Textbox
            label="New Question"
            placeholder="Add Question here"
            value={newQuestionValue}
            onChange={handleOnQuestionChange}
          />
          <Textbox
            label="Choices"
            placeholder="Choices seperated by comma"
            value={newQuestionChoiceValue}
            onChange={handleOnQuestionChoiceChange}
          />
          <AddQuestion.Button
            disabled={disableAddQuestionBtn}
            onClick={handleOnSaveQuestionClick}
          >
            Add Question
          </AddQuestion.Button>
        </StyledAddQuestionContainer>
      </PageContainer>
      <PageContainer title="Questions">
        <StyledContainer>
          {isLoading && (
            <>
              <Question.QuestionItemSkeleton />
              <Question.QuestionItemSkeleton />
              <Question.QuestionItemSkeleton />
              <Question.QuestionItemSkeleton />
            </>
          )}
          {!isLoading && (
            <>
              {questions.map(q => (
                <Question.QuestionItem
                  key={q.id}
                  questionItem={q}
                  onClick={handleOnQuestionClick(q.id)}
                />
              ))}
            </>
          )}
        </StyledContainer>
      </PageContainer>
    </>
  );
};

Questions.defaultProps = {
  getQuestions: () => {},
  saveQuestion: () => {}
};

Questions.propTypes = {
  getQuestions: PropTypes.func,
  saveQuestion: PropTypes.func,
  questions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      question: PropTypes.string.isRequired,
      publishedAt: PropTypes.string.isRequired,
      noOfChoices: PropTypes.number.isRequired
    })
  ).isRequired,
  isLoading: PropTypes.bool.isRequired
};

export default Questions;
