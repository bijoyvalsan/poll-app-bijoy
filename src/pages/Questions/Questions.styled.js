import styled from "styled-components";

export const StyledContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const StyledAddQuestionContainer = styled.div`
  display: flex;
  flex-direction: column;
  max-width: 600px;
  margin: 12px;
  box-sizing: border-box;
`;
