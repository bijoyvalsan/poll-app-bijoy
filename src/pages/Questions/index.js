const Questions = () => ('Hello Question')

export default Questions;
