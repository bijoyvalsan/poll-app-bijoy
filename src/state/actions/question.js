import { fetchQuestionById, saveVote } from "../../services/questionServices";

export const QUESTION_SINGLE_FETCH = "QUESTION_SINGLE_FETCH";
export const QUESTION_SINGLE_FETCH_SUCCSSES = "QUESTION_SINGLE_FETCH_SUCCSSES";
export const QUESTION_SINGLE_FETCH_ERROR = "QUESTION_SINGLE_FETCH_ERROR";

export const QUESTION_SINGLE_VOTE = "QUESTION_SINGLE_VOTE";
export const QUESTION_SINGLE_VOTE_SUCCSSES = "QUESTION_SINGLE_VOTE_SUCCSSES";
export const QUESTION_SINGLE_VOTE_ERROR = "QUESTION_SINGLE_VOTE_ERROR";

export const getQuestionById = urlPath => dispatch => {
  dispatch({
    type: QUESTION_SINGLE_FETCH
  });

  return fetchQuestionById(urlPath)
    .then(payload => {
      dispatch({
        type: QUESTION_SINGLE_FETCH_SUCCSSES,
        payload
      });
    })
    .catch(() => {
      dispatch({
        type: QUESTION_SINGLE_FETCH_ERROR
      });
    });
};

export const saveVoteForQuestion = urlPath => dispatch => {
  dispatch({
    type: QUESTION_SINGLE_VOTE
  });

  return saveVote(urlPath)
    .then(payload => {
      dispatch({
        type: QUESTION_SINGLE_VOTE_SUCCSSES,
        payload
      });
    })
    .catch(() => {
      dispatch({
        type: QUESTION_SINGLE_VOTE_ERROR
      });
    });
};
