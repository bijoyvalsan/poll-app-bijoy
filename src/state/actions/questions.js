import { fetchQuestions } from "../../services/questionServices";

export const QUESTIONS_FETCH = "QUESTIONS_FETCH";
export const QUESTIONS_FETCH_SUCCSSES = "QUESTIONS_FETCH_SUCCSSES";
export const QUESTIONS_FETCH_ERROR = "QUESTIONS_FETCH_ERROR";

export const getQuestions = () => dispatch => {
  dispatch({
    type: QUESTIONS_FETCH
  });

  fetchQuestions()
    .then(payload => {
      dispatch({
        type: QUESTIONS_FETCH_SUCCSSES,
        payload
      });
    })
    .catch(() => {
      dispatch({
        type: QUESTIONS_FETCH_ERROR
      });
    });
};
