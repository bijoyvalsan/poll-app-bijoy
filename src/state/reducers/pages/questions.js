import {
  QUESTIONS_FETCH,
  QUESTIONS_FETCH_ERROR,
  QUESTIONS_FETCH_SUCCSSES
} from "../../actions/questions";

export const initialState = {
  isLoading: true,
  hasError: false,
  questionCollection: []
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case QUESTIONS_FETCH: {
      return {
        ...state,
        isLoading: true,
        hasError: false
      };
    }
    case QUESTIONS_FETCH_SUCCSSES: {
      const data = action.payload;
      return {
        ...state,
        isLoading: false,
        hasError: false,
        questionCollection: data
      };
    }
    case QUESTIONS_FETCH_ERROR: {
      return {
        ...state,
        isLoading: false,
        hasError: true
      };
    }
    default:
      return state;
  }
};

export default reducer;
