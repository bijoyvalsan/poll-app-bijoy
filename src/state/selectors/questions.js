import { formatDate } from "../../helpers/date";

export const getQuestionsForListPage = state => {
  return state.pages.questions.questionCollection.map(q => ({
    question: q.question,
    publishedAt: formatDate(q.published_at),
    noOfChoices: q.choices.length,
    id: q.url
  }));
};

export const isQuestionsPageLoading = state => {
  return state.pages.questions.isLoading;
};
