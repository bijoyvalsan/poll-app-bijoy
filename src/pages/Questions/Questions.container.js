import { connect } from "react-redux";

import Questions from "./Questions";

import { getQuestions, saveQuestion } from "../../state/actions/questions";
import {
  getQuestionsForListPage,
  isQuestionsPageLoading
} from "../../state/selectors/questions";

const mapStateToProps = state => ({
  questions: getQuestionsForListPage(state),
  isLoading: isQuestionsPageLoading(state)
});

const mapDispatchToProps = {
  getQuestions,
  saveQuestion
};

export default connect(mapStateToProps, mapDispatchToProps)(Questions);
