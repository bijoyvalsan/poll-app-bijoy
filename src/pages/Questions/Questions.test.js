import React from "react";
import { render, waitFor, screen, fireEvent } from "@testing-library/react";

import Questions from "./Questions";
import * as QuestionsActions from "../../state/actions/questions";

describe("Questions Component", () => {
  let getQuestionsSpy;
  let saveQuestionSpy;

  beforeEach(() => {
    getQuestionsSpy = jest.spyOn(QuestionsActions, "getQuestions");
    saveQuestionSpy = jest.spyOn(QuestionsActions, "saveQuestion");
  });

  afterEach(() => {
    getQuestionsSpy.mockRestore();
    saveQuestionSpy.mockRestore();
  });

  it("should render with default Questions styles", () => {
    const mockData = {
      getQuestions: jest.fn(),
      saveQuestion: jest.fn(),
      isLoading: false,
      questions: [
        {
          id: "1",
          question: "Who am i ?",
          publishedAt: "10/20/2021",
          noOfChoices: 3
        },
        {
          id: "2",
          question: "Who am i ?",
          publishedAt: "10/20/2021",
          noOfChoices: 3
        }
      ]
    };
    const { asFragment } = render(<Questions {...mockData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with default QuestionsSkelton styles", () => {
    const mockData = {
      getQuestions: jest.fn(),
      saveQuestion: jest.fn(),
      isLoading: true,
      questions: [
        {
          id: "1",
          question: "Who am i ?",
          publishedAt: "10/20/2021",
          noOfChoices: 3
        },
        {
          id: "2",
          question: "Who am i ?",
          publishedAt: "10/20/2021",
          noOfChoices: 3
        }
      ]
    };
    const { asFragment } = render(<Questions {...mockData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should Invoke getQuestions and saveQuestion", async () => {
    const mockData = {
      getQuestions: getQuestionsSpy,
      saveQuestion: saveQuestionSpy,
      isLoading: false,
      questions: [
        {
          id: "1",
          question: "Who am i ?",
          publishedAt: "10/20/2021",
          noOfChoices: 3
        },
        {
          id: "2",
          question: "Who am i ?",
          publishedAt: "10/20/2021",
          noOfChoices: 3
        }
      ]
    };

    await waitFor(() => {
      render(<Questions {...mockData} />);
    });
    expect(getQuestionsSpy).toBeCalled();

    const addQuestionBtn = screen.queryByTestId("add-question-btn");
    const txtQuestion = screen.queryByTestId("txt-question");
    const txtQuestionChoice = screen.queryByTestId("txt-question-choice");

    expect(addQuestionBtn.getAttribute("disabled")).toBe("");

    fireEvent.change(txtQuestion, { target: { value: "test question?" } });
    fireEvent.change(txtQuestionChoice, { target: { value: "1,2,3" } });
    fireEvent.click(addQuestionBtn);

    await waitFor(
      () => {
        expect(addQuestionBtn.getAttribute("disabled")).toBeNull();
        expect(saveQuestionSpy).toBeCalled();
      },
      { timeout: 1000 }
    );
  });
});
