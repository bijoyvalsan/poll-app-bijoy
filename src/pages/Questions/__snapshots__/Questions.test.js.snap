// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions Component should render with default Questions styles 1`] = `
<DocumentFragment>
  <div
    class="sc-bqGGPW bvGpDM"
  >
    <div
      class="sc-ksluID eHYyPg"
    >
      <h2
        class="sc-hBMUJo igLrVS"
        data-testid="page-container-title"
      >
        Add Question
      </h2>
    </div>
    <div
      class="sc-iemWCZ jDoqWm"
    >
      <p
        class="sc-fnVZcZ idQmOh"
      >
        New Question
      </p>
      <div
        class="sc-fFSPTT fktQCA"
      >
        <input
          aria-label="Add Question here"
          data-testid="txt-question"
          id="txt-question"
          placeholder="Add Question here"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
      <p
        class="sc-fnVZcZ idQmOh"
      >
        Choices
      </p>
      <div
        class="sc-fFSPTT fktQCA"
      >
        <input
          aria-label="Choices seperated by comma"
          data-testid="txt-question-choice"
          id="txt-question-choice"
          placeholder="Choices seperated by comma"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
      <button
        class="sc-iCoGMd jtHMFE"
        data-testid="add-question-btn"
        disabled=""
      >
        Add Question
      </button>
    </div>
  </div>
  <div
    class="sc-bqGGPW bvGpDM"
  >
    <div
      class="sc-ksluID eHYyPg"
    >
      <h2
        class="sc-hBMUJo igLrVS"
        data-testid="page-container-title"
      >
        Questions
      </h2>
    </div>
    <div
      class="sc-bkbkJK iEoYTE"
    >
      <div
        class="sc-gtsrHT cgfNFa"
      >
        <p
          class="sc-eCApnc dejVSI"
        >
          Who am i ?
        </p>
        <span
          class="sc-jSFjdj jwWkJz"
        >
          Published: 10/20/2021
        </span>
        <span
          class="sc-gKAaRy hZEQAb"
        >
          No. Choices: 3
        </span>
      </div>
      <div
        class="sc-gtsrHT cgfNFa"
      >
        <p
          class="sc-eCApnc dejVSI"
        >
          Who am i ?
        </p>
        <span
          class="sc-jSFjdj jwWkJz"
        >
          Published: 10/20/2021
        </span>
        <span
          class="sc-gKAaRy hZEQAb"
        >
          No. Choices: 3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Questions Component should render with default QuestionsSkelton styles 1`] = `
<DocumentFragment>
  <div
    class="sc-bqGGPW bvGpDM"
  >
    <div
      class="sc-ksluID eHYyPg"
    >
      <h2
        class="sc-hBMUJo igLrVS"
        data-testid="page-container-title"
      >
        Add Question
      </h2>
    </div>
    <div
      class="sc-iemWCZ jDoqWm"
    >
      <p
        class="sc-fnVZcZ idQmOh"
      >
        New Question
      </p>
      <div
        class="sc-fFSPTT fktQCA"
      >
        <input
          aria-label="Add Question here"
          data-testid="txt-question"
          id="txt-question"
          placeholder="Add Question here"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
      <p
        class="sc-fnVZcZ idQmOh"
      >
        Choices
      </p>
      <div
        class="sc-fFSPTT fktQCA"
      >
        <input
          aria-label="Choices seperated by comma"
          data-testid="txt-question-choice"
          id="txt-question-choice"
          placeholder="Choices seperated by comma"
          tabindex="0"
          type="text"
          value=""
        />
      </div>
      <button
        class="sc-iCoGMd jtHMFE"
        data-testid="add-question-btn"
        disabled=""
      >
        Add Question
      </button>
    </div>
  </div>
  <div
    class="sc-bqGGPW bvGpDM"
  >
    <div
      class="sc-ksluID eHYyPg"
    >
      <h2
        class="sc-hBMUJo igLrVS"
        data-testid="page-container-title"
      >
        Questions
      </h2>
    </div>
    <div
      class="sc-bkbkJK iEoYTE"
    >
      <div
        class="sc-gtsrHT iCUPGs"
      >
        <div
          class="sc-dlnjwi sc-hKFxyN jkEfMo bGfFQW"
          width="60%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
      </div>
      <div
        class="sc-gtsrHT iCUPGs"
      >
        <div
          class="sc-dlnjwi sc-hKFxyN jkEfMo bGfFQW"
          width="60%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
      </div>
      <div
        class="sc-gtsrHT iCUPGs"
      >
        <div
          class="sc-dlnjwi sc-hKFxyN jkEfMo bGfFQW"
          width="60%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
      </div>
      <div
        class="sc-gtsrHT iCUPGs"
      >
        <div
          class="sc-dlnjwi sc-hKFxyN jkEfMo bGfFQW"
          width="60%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
        <div
          class="sc-dlnjwi sc-hKFxyN hEEJwU eVeVZB"
          width="90%"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
